### ISSUE TITLE
bootstrap: LLVM build disregards LDFLAGS

### ISSUE THREAD

**Issue by jonhoo** [rust-lang/rust#93880]
When the bootstrap code builds LLVM, it sometimes sets the CMake variable `CMAKE_EXE_LINKER_FLAGS`. For example, when instructed to also build the LLVM tools:

https://github.com/rust-lang/rust/blob/502d6aa47b4118fea1e326529e71b25a99b0d6c5/src/bootstrap/native.rs#L242-L250

When `CMAKE_EXE_LINKER_FLAGS` is unset, CMake [initializes it to `LDFLAGS`](https://gitlab.kitware.com/cmake/cmake/-/issues/18422#note_471785), but it does not consider `LDFLAGS` if `CMAKE_EXE_LINKER_FLAGS` is set. This means that Rust's LLVM bootstrap will _sometimes_ respect (potentially required) `LDFLAGS` and sometimes not depending on a number of unexpected factors, such as whether llvm tools are also being built, whether we're building for a non-FreeBSD RISC-V target, or if we're building for `powerpc-unknown-freebsd`. This, in turn, makes it difficult to bootstrap Rust in situations where additional linker flags are _required_, such as if dependencies are not stored in standard system-wide location such as `/usr/lib`.

I propose that anywhere we set `CMAKE_EXE_LINKER_FLAGS`, we append `LDFLAGS` from the environment to the value to retain parity with the behavior when the configuration isn't set.

**Comment by Mark-Simulacrum** [comment]
That makes sense to me, would be happy to accept a PR. It may also make sense to ask/file a similar bug against llvm recommending the logic to append always happen, since it seems like better behavior to me.

**Comment by jonhoo** [comment]
I don't think this is really a bug in llvm; this is a [standard CMAKE flag](https://cmake.org/cmake/help/latest/variable/CMAKE_EXE_LINKER_FLAGS.html), and the observed behavior is expected behavior; changing it would be a breaking change to cmake sadly.

I'd be happy to push a PR, though looking through the code raised a few other questions for me:

---

First, when `llvm.ldflags` is set, the current implementation sets the `CMAKE_*_LINKER_FLAGS` variables unilaterally
https://github.com/rust-lang/rust/blob/502d6aa47b4118fea1e326529e71b25a99b0d6c5/src/bootstrap/native.rs#L553-L557

And since `configure_cmake` is run _after_ all the logic that sets the linker flags based on other parts of the config, here:
https://github.com/rust-lang/rust/blob/502d6aa47b4118fea1e326529e71b25a99b0d6c5/src/bootstrap/native.rs#L354

It means that `llvm.ldflags` is currently an _override_, not an append. Do we want to keep that behavior, meaning that if `llvm.ldflags` is set in `config.toml`, `LDFLAGS` will _not_ be appended? Should we document that `llvm.ldflags` overrides any ldflags the bootstrap may want to set? Or should we change it to be appending?

---

Second, the current bootstrap logic sets `CMAKE_EXE_LINKER_FLAGS` multiple times:
https://github.com/rust-lang/rust/blob/502d6aa47b4118fea1e326529e71b25a99b0d6c5/src/bootstrap/native.rs#L247
https://github.com/rust-lang/rust/blob/502d6aa47b4118fea1e326529e71b25a99b0d6c5/src/bootstrap/native.rs#L263
https://github.com/rust-lang/rust/blob/502d6aa47b4118fea1e326529e71b25a99b0d6c5/src/bootstrap/native.rs#L312

Most of those are mutually exclusive, but when they're not, the [later setting takes precedence](https://gitlab.kitware.com/cmake/cmake/-/issues/21707). Is this intentional, or should the logic be fixed so that `ldflags` is added to, and then set in a single place at the end?

Anecdotally, we handle `CFLAGS` by appending:
https://github.com/rust-lang/rust/blob/502d6aa47b4118fea1e326529e71b25a99b0d6c5/src/bootstrap/native.rs#L522-L525
and same with `CXXFLAGS`:
https://github.com/rust-lang/rust/blob/502d6aa47b4118fea1e326529e71b25a99b0d6c5/src/bootstrap/native.rs#L530-L536

---

Third, it turns out we have a similar problem with `CFLAGS` and `CXXFLAGS`, except those we _always_ override, so `CFLAGS` and `CXXFLAGS` from the environment is _never_ respected. Should that be fixed too?

**Comment by Mark-Simulacrum** [comment]
I'm pretty sure that most of this behavior is unintentional and can largely be explained by incremental changes without really looking above/below in the file to make sure things aren't being set twice etc. I'm not particularly worried about breaking changes in bootstrap (e.g., if we start respecting LDFLAGS or CXXFLAGS etc), particularly when we're moving to what seems like relatively cleaner model.

In general happy to see a PR that goes through and cleans the logic up.

**Comment by jonhoo** [comment]
PR: https://github.com/rust-lang/rust/pull/93918
